### PR TITLE
Add nulltypes to parser

### PR DESCRIPTION
**Pull request by umerazad**
This change adds basic 4 nullable types to entity parser (JIRA-926/929). It includes a few bug fixes to existing code as well. This change also covers parts of the request path. Response path along with remaining Yarpc Connector changes will be covered in following CRs.

### CODE REVIEW

File: finder.go [hunk 1]
```diff
@@ -246,7 +248,7 @@ func tableFromStructType(structName string, structType *ast.StructType, packageP
 	return t, nil
 }
 
-func stringToDosaType(inType string, packagePrefix string) Type {
+func stringToDosaType(inType string) Type {
```
Comment (rkuris) [inline finder.go]: Removing this means that when someone imports dosa and renames the package, it won't work correctly. Please add a test entity with dosa renamed that has the new types.

File: entity_parser_test.go [hunk 2]
```diff
@@ -255,6 +302,54 @@ func TestUnsupportedType(t *testing.T) {
...
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "Nullable types are not allowed")
+}
+
+type NullFloat64PrimaryKeyType struct {
+	Entity          `dosa:"primaryKey=NullFloat64Type"`
+	NullFloat64Type NullFloat64
+}
+
+func TestNullFloat64PrimaryKeyType(t *testing.T) {
```
Comment (rkuris) [inline entity_parser_test.go]: Nit: these tests are all the same and could be done as one

File: finder_test.go [hunk 3]
```diff
@@ -54,8 +54,8 @@ func TestNonExistentDirectory(t *testing.T) {
 
 func TestParser(t *testing.T) {
 	entities, errs, err := FindEntities([]string{"."}, []string{})
-	assert.Equal(t, 13, len(entities), fmt.Sprintf("%s", entities))
-	assert.Equal(t, 14, len(errs), fmt.Sprintf("%v", errs))
+	assert.Equal(t, 15, len(entities), fmt.Sprintf("%s", entities))
+	assert.Equal(t, 17, len(errs), fmt.Sprintf("%v", errs))
```
Comment (rkuris) [inline finder_test.go]: I think you added 4 negative tests and one positive, so these do not add up correctly.

File: type.go [hunk 4]
```diff
@@ -106,7 +118,24 @@ func FromString(s string) Type {
...
+	case TNullBool.String():
+		return TNullBool
 	default:
 		return Invalid
 	}
 }
+
+func isNullableType(t Type) bool {
+	switch t {
+	case Invalid, TNullString, TNullFloat64, TNullInt64, TNullBool:
```
Comment (rkuris) [inline type.go]: Why is Invalid nullable?

File: finder.go (hunk 1 above)
Comment (umerazad) [inline finder.go]: Good catch. Fixed.

File: finder_test.go (hunk 3 above)
Comment (umerazad) [inline finder_test.go]: Negative tests are related to primary keys. The positive cases are all covered by the AllTypes test.

File: type.go (hunk 4 above)
Comment (umerazad) [inline type.go]: I've renamed the function to isValidPrimaryKeyType.

File: .gitignore [hunk 5]
```diff
@@ -81,6 +81,8 @@ vendor/
 # JIRA plugin
 atlassian-ide-plugin.xml
 
+# VSCode
+.vscode/
```
Comment (rkuris) [inline .gitignore]: This is the generated portion of this file. Best bet is to regenerate it with [gitignore](https://www.gitignore.io/api/go,tags,intellij+iml,visualstudiocode) which adds visualstudiocode support, or alternatively add .vscode to the beginning of this file, before the autogenerated parts.

File: cmd/dosa/schema_test.go [hunk 6]
```diff
@@ -284,7 +285,7 @@ func TestSchema_Upsert_Happy(t *testing.T) {
 				dl, ok := ctx.Deadline()
 				assert.True(t, ok)
 				assert.True(t, dl.After(time.Now()))
-				assert.Equal(t, 1, len(ed))
+				assert.Equal(t, 4, len(ed))
 				assert.Equal(t, "awesome_test_entity", ed[0].Name)
```
Comment (rkuris) [inline cmd/dosa/schema_test.go]: Isn't this now a floppy test? There is nothing in the contract that says that awesome_test_entity must be first in the returned list.

File: doc.go [hunk 7]
```diff
@@ -22,13 +22,13 @@
 //
 // Abstract
 //
-// :warning: DOSA is *BETA* software. It is not recommended for production use.
+// :warning: DOSA is BETA software. It is not recommended for production use.
```
Comment (rkuris) [inline doc.go]: not sure why this change happened, probably differing versions of docgen?

File: entity.go [hunk 8]
```diff
@@ -177,6 +177,27 @@ func (e *EntityDefinition) EnsureValid() error {
 		keyNamesSeen[c.Name] = struct{}{}
 	}
 
+	if err := e.ensureNonNullablePrimaryKeys(); err != nil {
+		return errors.Wrap(err, "Nullable types are not allowed as primary key(s)")
```
Comment (rkuris) [inline entity.go]: nit: not sure I'd wrap; you're not adding any additional info to the existing error. `primary key is of nullable type: "%q"` is enough of an error

File: type.go [hunk 9]
```diff
@@ -106,7 +118,24 @@ func FromString(s string) Type {
...
+	case TNullBool.String():
+		return TNullBool
 	default:
 		return Invalid
 	}
 }
+
+func isValidPrimaryKeyType(t Type) bool {
+	switch t {
+	case Invalid, TNullString, TNullFloat64, TNullInt64, TNullBool:
```
Comment (rkuris) [inline type.go]: This looks backwards. `isValidPrimaryKeyType` of `TNullFloat` returns `true` but `TNullFloat` is not a valid primary key type. Perhaps you meant `isInvalidPrimaryKeyType`?

File: testentity/named_import_testentity.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package testentity
+
+import (
+	"time"
+
+	// Using a named import is the key change in this file. This is
+	// used to test the entity parser against named dosa imports.
```
Comment (rkuris) [inline testentity/named_import_testentity.go]: 👍

File: scan_test.go [hunk 11]
```diff
@@ -30,16 +30,20 @@ import (
...
-	dosa.Entity `dosa:"primaryKey=BoolType"`
-	BoolType    bool
-	Int32Type   int32
-	Int64Type   int64
-	DoubleType  float64
-	StringType  string
-	BlobType    []byte
-	TimeType    time.Time
-	UUIDType    dosa.UUID
+type AllTypesScanTestEntity struct {
```
Comment (rkuris) [inline scan_test.go]: Why was this renamed? Seems like having a duplicate name in a different package is a good test.

File: .gitignore (hunk 5 above)
Comment (umerazad) [inline .gitignore]: Fixed

File: cmd/dosa/schema_test.go (hunk 6 above)
Comment (umerazad) [inline cmd/dosa/schema_test.go]: It wasn't floppy because the entity names were same. But I've fixed it to be more explicit and resilient.

File: doc.go (hunk 7 above)
Comment (umerazad) [inline doc.go]: Not sure. I'll look into it separately.

File: entity_parser_test.go [hunk 12]
```diff
@@ -198,23 +198,27 @@ func TestMissingAnnotation(t *testing.T) {
...
+	Entity          `dosa:"primaryKey=BoolType"`
+	BoolType        bool
+	Int32Type       int32
+	Int64Type       int64
+	DoubleType      float64
+	StringType      string
+	BlobType        []byte
+	TimeType        time.Time
+	UUIDType        UUID
+	NullStringType  NullString
```
Comment (junchaowu) [inline entity_parser_test.go]: so we only support these four null type?

File: entity.go (hunk 8 above)
Comment (umerazad) [inline entity.go]: I've removed it.

File: type.go (hunk 9 above)
Comment (umerazad) [inline type.go]: Ooops. Good catch. Fixed.

File: scan_test.go (hunk 11 above)
Comment (umerazad) [inline scan_test.go]: It took me good couple of hours to debug an issue in the walker because initially I only added the new types in AllTypes structure but apparently there were two AllTypes structures and TestParser test in (finder_test.go) would fail at random iterations. It was very confusing and painful to debug. This particular definition in scan_test.go isn't used anywhere else but it needed to be exact same structure as AllTypes defined in entity_parser_test.go.

File: entity_parser_test.go (hunk 12 above)
Comment (umerazad) [inline entity_parser_test.go]: NullTime and NullGUID will be supported too.
